python/ray/train/callbacks/profile.py

```diff
@@ -0,0 +1,93 @@
+import logging
+import pathlib
+from typing import List, Dict, Optional, Union
+
+from torch.profiler import profile, record_function, schedule
+
+from ray import train
+from ray.train import TrainingCallback
+from ray.train.callbacks.logging import TrainingLogdirMixin
+from ray.train.callbacks.results_prepocessors import KeysResultsPreprocessor, \
+    WorkersResultsPreprocessor
+from ray.train.constants import PROFILER_KEY
+
+logger = logging.getLogger(__name__)
+
+WORKER_TRACE_DIR = "worker_traces"
+DRIVER_TRACE_DIR = "pytorch_profiler"
+
+
+class TorchWorkerProfiler():
+    def __init__(self, trace_dir: Optional[str] = None):
+        self.profiler_traces = []
+        self.recorded_functions = {}
+
+        trace_dir = trace_dir or WORKER_TRACE_DIR
+        self.trace_dir = pathlib.Path(trace_dir)
+        self.trace_dir.mkdir(parents=True, exist_ok=True)
+
+        self.p = profile(
+            activities=[],  # default activities
+            schedule=schedule(wait=0, warmup=0, active=1),
+            on_trace_ready=self.trace_handler)
```

Comment:
We should make it configurable latter down the line

Reply:
Yeah let me see if there's a way so the user can directly use the profiler. If not I was thinking about taking a `profile_args: Dict` arg and passing it to `profile(**profile_args)`.

Reply:
Is there any hard reason for us to provide a wrapper in the first place? It's more code to change for the user if they are already using the pytorch profiler. We can provide our own trace handler instead, maybe?

Reply:
No hard reason, originally I wanted to create a new class so that it encompasses all the methods a user might need, but to your point configurability or conversion from existing usage would be strange. Taking another look I can slim it down so that the class just contains `__init__`, `trace_handler`, and `get_and_clear_profile_traces`, and maybe even further cleaned up depending how we set up the flow to push the traces into `train.report`

Reply:
Or just make the path to the trace json an argument in `TorchTensorboardProfilerCallback`

Reply:
> Or just make the path to the trace json an argument in `TorchTensorboardProfilerCallback` - that would be the simplest way imo

Ah the problem with this is that the Callback is on the driver while the Profiler is on the worker.

Maybe yet another way would be to not use Callbacks for this and just `rsync` the traces?

Reply:
Ah, right - I'd just have some small utility method to load the trace into results then. Ideally, that would mean there is only a 1-2 line change in order to use existing profiler code with Train.

Reply:
For now I simplified the logic based on this so that the user defines their own profiler. 

1. `TorchWorkerProfiler` is used to keep track of state (`trace_handler` does not return anything).
2. `trace_handler` writes the trace and simply stores the filename in `profiler_trace_filenames`
3. `get_and_clear_profile_traces` will then read all the trace files and return a datamap that is passed to `train.report()` and consumed by `TorchTensorboardProfilerCallback`.

In regards to slimming it down further, I think the benefit right now with the provided methods is that the user does not need to manually construct or handle filenames/paths. Technically the user can still handle their traces however they want, and utilize the Callback as long as they conform to the `List[(filename, text)]` format